### PR TITLE
[PATCH CLOUD-DEV v5] Modular pktio ops

### PR DESCRIPTION
**Pull request by heyi-linaro**
Part one (3 commits): Apply modular framework to the pktio ops subsystem
- Previously we register pktio ops in a static array pktio_if_ops[]; 
- This converts it into a modular and extensible manner.

Part two (7 commits): Minor code re-factory on each pktio ops implementation
- Cleanup the odp_packet_io_internal.h header file
- Move each impl specific data structure into a dedicated header file
- Consistent the header file names and data structure names.
- Dpdk pktio ops demonstrates to hide the impl private data locally.

Part three Makefiles re-organization
- Read #127 to learn Dmitry's working on build system: no conflict or duplicate works
- Wrote a Makefile framework which allows new subsystem and module writers to easily add their works by providing brief descriptions only, and Makefile rules and targets etc will be automatically generated.
- In internal review process

### CODE REVIEW

File: platform/linux-generic/pktio/subsystem.c [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+ODP_SUBSYSTEM_CONSTRUCTOR(pktio_ops)
+{
+	odp_subsystem_constructor(pktio_ops);
+
+	/* Further initialization per subsystem */
+
+#ifdef ODP_PKTIO_DPDK
+	enable_link_dpdk_pktio_ops = 1;
+#endif
```
Comment (GBalakrishna) [inline platform/linux-generic/pktio/subsystem.c]: I feel we don't need PKTIO_DPDK anylonger as we now already have it from linux-dpdk. I have sent a PR [https://github.com/Linaro/odp/pull/140](url) with necessary updates.

**Comment by Bill-Fischofer-Linaro** [comment]
Note that this PR appears to have broken the linux-dpdk build, as reported by Travis. We can discuss how best to resolve this during tomorrow's call.

File: platform/linux-generic/pktio/subsystem.c (hunk 1 above)
Comment (heyi-linaro) [inline platform/linux-generic/pktio/subsystem.c]: Yes, Krishna, as I replied in another email, I suggest an approach that add workable solutions first and after then remove obsolete code.

With PR#140 it looks likes that we removed linux-generic dpdk pktio and lost dpdk pktio at all, now wait for a new design which seems still in discussion.

I'll comment these in your PR as a review suggestion.

Comment (GBalakrishna) [inline platform/linux-generic/pktio/subsystem.c]: We have the working solution currently in linux-dpdk. Question is, why we still want to have it in linux-generic ?

**Comment by heyi-linaro** [comment]
While work on 3rd part of Makefile re-organization, I met a difficulty and very like to share with you, and maybe you have more experience with automake and can help provide suggestions:

I'm trying to simplify subsystem and module writers effort to add their code into Makefile infra, I create a Makefile-paths.include, allow writers to do only two steps:

- 1st add the new subsystem into variable SUBSYSTEMS
SUBSYSTEMS += new

- 2nd list the modules (implementations) of this subsystem, and can specify the selection of one module as active(default) by appending an ":active" suffix
__subsystem_new_MODULES = one:active two three four

And that's all, I wrote a Makefile-subsystems.include which automatically generate all targets and rules for build the subsystem and modules based on above description.

Unfortunately it does not work with automake, the works are pushed:
- https://github.com/heyi-linaro/odp/tree/modular-pktio-ops-Makefiles

In the branch you will find the platform/linux-generic/Makefile-paths.inc and Makefile-subsystems.inc described above.

The reason I figured currently is because automake statically analyze Makefile.am and generate Makefile.in, without run any of the dynamic expansions included in Makefile-subsystems.inc. But what I needed is exactly that.

I'll continue back next week, sorry today will not be able to send out internal review.

Thanks and best regards, Yi

File: platform/linux-generic/pktio/dpdk.c [hunk 2]
```diff
@@ -1290,30 +1290,46 @@ static int dpdk_stats_reset(pktio_entry_t *pktio_entry)
 	return 0;
 }
 
-const pktio_if_ops_t dpdk_pktio_ops = {
-	.name = "dpdk",
-	.init_global = dpdk_pktio_init_global,
-	.init_local = dpdk_pktio_init_local,
-	.term = NULL,
+static pktio_ops_module_t dpdk_pktio_ops = {
```
Comment (joseppc) [inline platform/linux-generic/pktio/dpdk.c]: Can we find a solution so these structures are kept as const? If I get this correctly, apart from the linked list embedded in the "base class", there is nothing that should be mutable here, right?

File: platform/linux-generic/pktio/pcap.c [hunk 3]
```diff
@@ -51,7 +51,8 @@ static const char pcap_mac[] = {0x02, 0xe9, 0x34, 0x80, 0x73, 0x04};
 
 static int pcapif_stats_reset(pktio_entry_t *pktio_entry);
 
-static int _pcapif_parse_devname(pkt_pcap_t *pcap, const char *devname)
+static int _pcapif_parse_devname(
+	pktio_ops_pcap_data_t *pcap, const char *devname)
```
Comment (joseppc) [inline platform/linux-generic/pktio/pcap.c]: indentation style

File: platform/linux-generic/pktio/pcap.c [hunk 4]
```diff
@@ -163,7 +165,8 @@ static int pcapif_init(odp_pktio_t id ODP_UNUSED, pktio_entry_t *pktio_entry,
 
 static int pcapif_close(pktio_entry_t *pktio_entry)
 {
-	pkt_pcap_t *pcap = &pktio_entry->s.pkt_pcap;
+	pktio_ops_pcap_data_t *pcap =
+		&pktio_entry->ops_data(pcap);
```
Comment (joseppc) [inline platform/linux-generic/pktio/pcap.c]: declaration fins in one line (there are others in this file too)

**Comment by heyi-linaro** [comment]
Part three Makefiles re-organization

Read #127 to learn Dmitry's working on build system: no conflict or duplicate works
Wrote a Makefile framework which allows new subsystem and module writers to easily add their works by providing brief descriptions only, and Makefile rules and targets etc will be automatically generated.
In internal review process

File: platform/linux-generic/include/odp_pktio_ops_subsystem.h [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+#include <odp_module.h>
+#include <odp/api/packet_io.h>
+
+/* ODP packet IO operations subsystem declaration */
+ODP_SUBSYSTEM_DECLARE(pktio_ops);
+
+/* Subsystem APIs declarations */
+ODP_SUBSYSTEM_API(pktio_ops, int, open, odp_pktio_t,
+		  pktio_entry_t *, const char *, odp_pool_t);
```
[the remark this answers is not in the document]
Comment (heyi-linaro) [reply platform/linux-generic/include/odp_pktio_ops_subsystem.h]: Yes, here we only use the ODP_SUBSYSTEM_API() macro to typedef prototype function pointers and leave the prototype declarations no defines.

I suggest we accept this as is and consider how to handle this kind of tasks more properly in terms of autogen. as described in https://github.com/heyi-linaro/pure-interface/issues/6

File: platform/linux-generic/pktio/subsystem.c [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+extern int enable_link_tap_pktio_ops;
+
+ODP_SUBSYSTEM_CONSTRUCTOR(pktio_ops)
+{
+	odp_subsystem_constructor(pktio_ops);
+
+	/* Further initialization per subsystem */
+
+#ifdef ODP_PKTIO_DPDK
+	enable_link_dpdk_pktio_ops = 1;
```
[the remark this answers is not in the document]
Comment (heyi-linaro) [reply platform/linux-generic/pktio/subsystem.c]: This is temporary variables to make sure these modules are linked into ODP library.

Because the Makefiles have not been re-organized to build modules each as real independent module (.a or .so) and either link them with --whole-archive or --no-as-needed flags, this is necessary for current Makefiles.

File: platform/linux-generic/pktio/dpdk.c (hunk 2 above)
Comment (heyi-linaro) [reply platform/linux-generic/pktio/dpdk.c]: Yes, I'm thinking how to achieve this. Can it be later patch improvements.

File: platform/linux-generic/pktio/pcap.c (hunk 4 above)
Comment (heyi-linaro) [reply platform/linux-generic/pktio/pcap.c]: Run out of 80 characters?

File: platform/linux-generic/pktio/pcap.c (hunk 3 above)
Comment (heyi-linaro) [reply platform/linux-generic/pktio/pcap.c]: Run out of 80 characters?

**Comment by heyi-linaro** [comment]
@Bill-Fischofer-Linaro @joseppc @nagarahalli @GBalakrishna 

This PR has been fixed for both linux-dpdk build and checkpatch issues.
For those commits both I and Krishna contributed, I've added Signed-off-by stamps.

**Comment by bogdanPricope** [comment]
Reviewed-by: Bogdan Pricope <bogdan.pricope@linaro.org>

Push?

**Comment by joseppc** [comment]
Reviewed-by: Josep Puigdemont <josep.puigdemont@linaro.org>

File: platform/linux-generic/pktio/socket_mmap.c [hunk 7]
```diff
@@ -710,18 +713,22 @@ static int sock_mmap_init_global(void)
...
 
-const pktio_if_ops_t sock_mmap_pktio_ops = {
-	.name = "socket_mmap",
-	.print = NULL,
-	.init_global = sock_mmap_init_global,
-	.init_local = NULL,
-	.term = NULL,
+static pktio_ops_module_t socket_mmap_pktio_ops = {
+	.base = {
+		.name = "socket mmap",
```
Comment (Bill-Fischofer-Linaro) [inline platform/linux-generic/pktio/socket_mmap.c]: Again you've changed the interface name here from `socket_mmap` to `socket mmap`.  Keep the names unchanged as part of this PR to avoid application impact.

File: platform/linux-generic/pktio/socket_mmap.c [hunk 8]
```diff
@@ -730,9 +737,20 @@ const pktio_if_ops_t sock_mmap_pktio_ops = {
...
+{
+	odp_module_constructor(&socket_mmap_pktio_ops);
+
+	odp_subsystem_register_module(pktio_ops, &socket_mmap_pktio_ops);
+}
+
+/* Temporary variable to enable link this module,
+ * will remove in Makefile scheme changes.
+ */
+int enable_link_socket_mmap_pktio_ops = 0;
```
Comment (Bill-Fischofer-Linaro) [inline platform/linux-generic/pktio/socket_mmap.c]: Same "to do" comment.

File: platform/linux-generic/pktio/tap.c [hunk 9]
```diff
@@ -373,24 +372,44 @@ static int tap_capability(pktio_entry_t *pktio_entry ODP_UNUSED,
...
+{
+	odp_module_constructor(&tap_pktio_ops);
+
+	odp_subsystem_register_module(pktio_ops, &tap_pktio_ops);
+}
+
+/* Temporary variable to enable link this module,
+ * will remove in Makefile scheme changes.
+ */
+int enable_link_tap_pktio_ops = 0;
```
Comment (Bill-Fischofer-Linaro) [inline platform/linux-generic/pktio/tap.c]: Same "to do" comment.

File: platform/linux-dpdk/pktio/dpdk.c [hunk 10]
```diff
@@ -664,34 +701,40 @@ static void stats_convert(struct rte_eth_stats *rte_stats,
...
+		__retrieve_op_data(pktio_entry);
+	rte_eth_stats_reset(pkt_dpdk->portid);
 	return 0;
 }
 
-const pktio_if_ops_t dpdk_pktio_ops = {
-	.name = "odp-dpdk",
+static pktio_ops_module_t dpdk_pktio_ops = {
+	.base = {
+		.name = "dpdk",
```
Comment (Bill-Fischofer-Linaro) [inline platform/linux-dpdk/pktio/dpdk.c]: Do not change interface names as part of this PR. Any name changes should be a separate PR where we can access application impact. In the case of DPDK devices this may be appropriate here, but not as part of this PR.

File: platform/linux-dpdk/pktio/dpdk.c [hunk 11]
```diff
@@ -712,3 +755,16 @@ const pktio_if_ops_t dpdk_pktio_ops = {
...
+	odp_module_constructor(&dpdk_pktio_ops);
+
+	odp_subsystem_register_module(pktio_ops, &dpdk_pktio_ops);
+}
+
+/* Temporary variable to enable link this module,
+ * will remove in Makefile scheme changes.
+ */
+int enable_link_dpdk_pktio_ops = 0;
+
```
[the remark this answers is not in the document]
Comment (heyi-linaro) [reply platform/linux-dpdk/pktio/dpdk.c]: Yes, created https://github.com/Linaro/odp/issues/184

File: platform/linux-dpdk/pktio/dpdk.c (hunk 10 above)
Comment (heyi-linaro) [reply platform/linux-dpdk/pktio/dpdk.c]: Yes, will fix all these.

File: platform/linux-generic/pktio/loopback.c [hunk 12]
```diff
@@ -251,18 +251,22 @@ static int loop_init_global(void)
...
-const pktio_if_ops_t loopback_pktio_ops = {
-	.name = "loop",
-	.print = NULL,
-	.init_global = loop_init_global,
-	.init_local = NULL,
-	.term = NULL,
+static pktio_ops_module_t loopback_pktio_ops = {
+	.base = {
+		.name = "loopback",
+		.init_local = NULL,
```
[the remark this answers is not in the document]
Comment (heyi-linaro) [reply platform/linux-generic/pktio/loopback.c]: Yes, will change it back to the original.

File: platform/linux-generic/pktio/loopback.c [hunk 13]
```diff
@@ -271,9 +275,20 @@ const pktio_if_ops_t loopback_pktio_ops = {
...
+{
+	odp_module_constructor(&loopback_pktio_ops);
+
+	odp_subsystem_register_module(pktio_ops, &loopback_pktio_ops);
+}
+
+/* Temporary variable to enable link this module,
+ * will remove in Makefile scheme changes.
+ */
+int enable_link_loopback_pktio_ops = 0;
```
[the remark this answers is not in the document]
Comment (heyi-linaro) [reply platform/linux-generic/pktio/loopback.c]: Yes, created #184

File: platform/linux-generic/pktio/pcap.c [hunk 14]
```diff
@@ -423,25 +429,44 @@ static int pcapif_init_global(void)
...
+{
+	odp_module_constructor(&pcap_pktio_ops);
+
+	odp_subsystem_register_module(pktio_ops, &pcap_pktio_ops);
+}
+
+/* Temporary variable to enable link this module,
+ * will remove in Makefile scheme changes.
+ */
+int enable_link_pcap_pktio_ops = 0;
```
[the remark this answers is not in the document]
Comment (heyi-linaro) [reply platform/linux-generic/pktio/pcap.c]: Yes, created #184

File: platform/linux-generic/pktio/socket.c [hunk 15]
```diff
@@ -873,9 +881,20 @@ const pktio_if_ops_t sock_mmsg_pktio_ops = {
...
+{
+	odp_module_constructor(&socket_pktio_ops);
+
+	odp_subsystem_register_module(pktio_ops, &socket_pktio_ops);
+}
+
+/* Temporary variable to enable link this module,
+ * will remove in Makefile scheme changes.
+ */
+int enable_link_socket_pktio_ops = 0;
```
[the remark this answers is not in the document]
Comment (heyi-linaro) [reply platform/linux-generic/pktio/socket.c]: Yes, created #184

File: platform/linux-generic/pktio/socket_mmap.c (hunk 7 above)
Comment (heyi-linaro) [reply platform/linux-generic/pktio/socket_mmap.c]: Will fix all these.

File: platform/linux-generic/pktio/socket_mmap.c (hunk 8 above)
Comment (heyi-linaro) [reply platform/linux-generic/pktio/socket_mmap.c]: Yes, created #184

File: platform/linux-generic/pktio/tap.c (hunk 9 above)
Comment (heyi-linaro) [reply platform/linux-generic/pktio/tap.c]: Yes, created #184

**Comment by heyi-linaro** [comment]
merged and closed.
